lib/commons/aria/lookup-table.js

```diff
@@ -2079,6 +2094,132 @@ lookupTable.role = {
 	}
 };
 
+// Source: https://www.w3.org/TR/html-aam-1.0/#element-mapping-table
+// Source: https://www.w3.org/TR/html-aria/
+lookupTable.implicitRole = {
+	a: vNode => {
+		return vNode.hasAttr('href') ? 'link' : null;
+	},
+	area: vNode => {
+		return vNode.hasAttr('href') ? 'link' : null;
+	},
+	article: 'article',
+	aside: 'complementary',
+	body: 'document',
+	button: 'button',
+	datalist: 'listbox',
+	dd: 'definition',
+	dfn: 'term',
+	details: 'group',
+	dialog: 'dialog',
+	dt: 'term',
+	fieldset: 'group',
+	figure: 'figure',
+	footer: vNode => {
+		const sectioningElement = closest(
+			vNode,
+			'article, aside, main, nav, section, [role=article], [role=complementary], [role=main], [role=navigation], [role=region]'
+		);
+
+		return !sectioningElement ? 'contentinfo' : null;
+	},
+	form: vNode => {
+		// accessible-text-virtual uses getRole, so we have to only
+		// use the name calculation steps we know apply to a form
+		// element
```

Comment:
This isn't going to be sustainable. For us to properly implement presentational role conflict resolution, titleText is going to have to call getRole(), not just on the `alwaysTitleElements` list, but on all elements with a title. This is also relying on us continuing to have arialabelledby-text not work recursively, which could potentially change.

I think we should instead figure out a way to ensure this can never get into an infinite loop. How do browsers do this? I think it's worth doing some reverse engineering to figure out how browsers dealt with this.